File: python/socketd/socketd/transport/core/codec/CodecDefault.py
from typing import Callable, Optional

from socketd.transport.core.Asserts import Asserts
from socketd.transport.core.Codec import Codec, CodecReader, CodecWriter
from socketd.transport.core.Costants import Constants
from socketd.transport.core.Flags import Flags
from socketd.transport.core.Frame import Frame
from socketd.transport.core.entity.MessageBuilder import MessageBuilder
from socketd.transport.core.entity.EntityDefault import EntityDefault
from socketd.transport.core.Config import Config
from socketd.transport.core.codec.Buffer import Buffer


class CodecDefault(Codec):
    def __init__(self, config: Config):
        self.config = config

    def write(self, frame: Frame, writerFactory: Callable[[int], CodecWriter]) -> CodecWriter:
        if frame.message() is None:
            # length (flag + int.bytes)
            _len = 2 * 4
            target: CodecWriter = writerFactory(_len)

            # length
            target.put_int(_len)
            # flag
            target.put_int(frame.flag())
            target.flush()

            return target
        else:
            # sid
            sidB: bytes = frame.message().sid().encode(self.config.get_charset())
            # event
            eventB: bytes = frame.message().event().encode(self.config.get_charset())
            # metaString
            metaStringB: bytes = frame.message().entity().meta_string().encode(self.config.get_charset())

            # length (len[int] + flag[int] + sid + event + metaString + data + \n*3)
            len1 = 4 + 4 + len(sidB) + len(eventB) + len(metaStringB) + frame.message().data_size() + 2 * 3

            Asserts.assert_size("sid", len(sidB), Constants.MAX_SIZE_SID)
            Asserts.assert_size("event", len(eventB), Constants.MAX_SIZE_EVENT)
            Asserts.assert_size("metaString", len(metaStringB), Constants.MAX_SIZE_META_STRING)
            Asserts.assert_size("data", frame.message().entity().data_size(), Constants.MAX_SIZE_DATA)

            target: CodecWriter = writerFactory(len1)

            # length
            target.put_int(len1)

            # flag
            target.put_int(frame.flag())

            # sid
            target.put_bytes(sidB)
            target.put_char(10) #'\n'

            # event
            target.put_bytes(eventB)
            target.put_char(10)

            # metaString
            target.put_bytes(metaStringB)
            target.put_char(10)

            # _data
            if frame.message().entity().data() is not None:
                target.put_bytes(frame.message().entity().data().getvalue())
            target.flush()

            return target

    def read(self, _reader: CodecReader) -> Frame | None:
        frameSize = _reader.get_int()

        if frameSize > (_reader.remaining() + 4):
            return None

        flag = _reader.get_int()  # 取前一位数据

        if frameSize == 8:
            # len + flag
            return Frame(Flags.of(flag), None)
        else:
            metaBufSize = min(Constants.MAX_SIZE_META_STRING, _reader.remaining())

            # 1. decode sid and event
            buf = Buffer(limit=metaBufSize)

            sid = self.decodeString(_reader, buf, Constants.MAX_SIZE_SID)

            event = self.decodeString(_reader, buf, Constants.MAX_SIZE_EVENT)

            metaString = self.decodeString(_reader, buf, Constants.MAX_SIZE_META_STRING)

            # 2. decode body
            dataRealSize = frameSize - _reader.position()
            data: Optional[bytearray] = None
            if dataRealSize > Constants.MAX_SIZE_DATA:
                # exceeded the limit, read and discard the bytes
                data = bytearray(Constants.MAX_SIZE_DATA)
                _reader.get_buffer().readinto(data)
                for i in range(dataRealSize - Constants.MAX_SIZE_DATA):
                    _reader.get_buffer().read()
            else:
                data = bytearray(_reader.get_buffer().read(dataRealSize))

            message = (MessageBuilder()
                       .flag(Flags.of(flag))
                       .sid(sid)
                       .event(event)
                       .entity(EntityDefault().data_set(data).meta_string_set(metaString))
                       .build())
            buf.close()
            _reader.close()
            return Frame(message.flag(), message)

    def decodeString(self, reader: CodecReader, buf: Buffer, maxLen: int) -> str:
        b = bytearray(reader.get_buffer().readline(maxLen).replace(b'\x00\n', b''))
        if buf.limit() < 1:
            return ""
        return b.decode(self.config.get_charset())

File: javascript/socketd/src/socketd/transport/core/codec/CodecDefault.ts
import type {Config} from "../Config";
import {Frame} from "../Frame";
import type {IoFunction} from "../Typealias";
import {StrUtils} from "../../../utils/StrUtils";
import {Asserts} from "../Asserts";
import {Constants} from "../Constants";
import {Flags} from "../Flags";
import {MessageBuilder} from "../Message"
import type {Codec, CodecReader, CodecWriter} from "../Codec";
import {EntityDefault} from "../entity/EntityDefault";

/**
 * 编解码器（基于 CodecReader,CodecWriter 接口编解）
 *
 * @author noear
 * @since 2.0
 */
export class CodecDefault implements Codec {
    private _config: Config;

    constructor(config: Config) {
        this._config = config;
    }

    /**
     * 解码写入
     *
     * @param frame         帧
     * @param targetFactory 目标工厂
     */
    write<T extends CodecWriter>(frame: Frame, targetFactory: IoFunction<number, T>): T {
        if (frame.message()) {
            //sid
            const sidB = StrUtils.strToBuf(frame.message()!.sid(), this._config.getCharset());
            //event
            const eventB = StrUtils.strToBuf(frame.message()!.event(), this._config.getCharset());
            //metaString
            const metaStringB = StrUtils.strToBuf(frame.message()!.metaString(), this._config.getCharset());

            //length (len[int] + flag[int] + sid + event + metaString + data + \n*3)
            const frameSize = 4 + 4 + sidB.byteLength + eventB.byteLength + metaStringB.byteLength + frame.message()!.dataSize() + 2 * 3;

            Asserts.assertSize("sid", sidB.byteLength, Constants.MAX_SIZE_SID);
            Asserts.assertSize("event", eventB.byteLength, Constants.MAX_SIZE_EVENT);
            Asserts.assertSize("metaString", metaStringB.byteLength, Constants.MAX_SIZE_META_STRING);
            Asserts.assertSize("data", frame.message()!.dataSize(), Constants.MAX_SIZE_DATA);

            const target = targetFactory(frameSize);

            //长度
            target.putInt(frameSize);

            //flag
            target.putInt(frame.flag());

            //sid
            target.putBytes(sidB);
            target.putChar('\n'.charCodeAt(0));

            //event
            target.putBytes(eventB);
            target.putChar('\n'.charCodeAt(0));

            //metaString
            target.putBytes(metaStringB);
            target.putChar('\n'.charCodeAt(0));

            //data
            target.putBytes(frame.message()!.data().getArray()!);

            target.flush();

            return target;
        } else {
            //length (len[int] + flag[int])
            const frameSize = 4 + 4;
            const target = targetFactory(frameSize);

            //长度
            target.putInt(frameSize);

            //flag
            target.putInt(frame.flag());
            target.flush();

            return target;
        }
    }

    /**
     * 编码读取
     *
     * @param buffer 缓冲
     */
    read(buffer: CodecReader): Frame | null { //=>Frame
        const frameSize = buffer.getInt();

        if (frameSize > (buffer.remaining() + 4)) {
            return null;
        }

        const flag = buffer.getInt();

        if (frameSize == 8) {
            //len[int] + flag[int]
            return new Frame(Flags.of(flag), null);
        } else {

            const metaBufSize = Math.min(Constants.MAX_SIZE_META_STRING, buffer.remaining());

            //1.解码 sid and event
            const buf = new ArrayBuffer(metaBufSize);

            //sid
            const sid = this.decodeString(buffer, buf, Constants.MAX_SIZE_SID);

            //event
            const event = this.decodeString(buffer, buf, Constants.MAX_SIZE_EVENT);

            //metaString
            const metaString = this.decodeString(buffer, buf, Constants.MAX_SIZE_META_STRING);

            //2.解码 body
            const dataRealSize = frameSize - buffer.position();
            let data: ArrayBuffer;
            if (dataRealSize > Constants.MAX_SIZE_DATA) {
                //超界了，空读。必须读，不然协议流会坏掉
                data = new ArrayBuffer(Constants.MAX_SIZE_DATA);
                buffer.getBytes(data, 0, Constants.MAX_SIZE_DATA);
                for (let i = dataRealSize - Constants.MAX_SIZE_DATA; i > 0; i--) {
                    buffer.getByte();
                }
            } else {
                data = new ArrayBuffer(dataRealSize);
                if (dataRealSize > 0) {
                    buffer.getBytes(data, 0, dataRealSize);
                }
            }

            //先 data , 后 metaString (避免 data 时修改元信息)
            const message = new MessageBuilder()
                .flag(Flags.of(flag))
                .sid(sid)
                .event(event)
                .entity(new EntityDefault().dataSet(data).metaStringSet(metaString))
                .build();

            return new Frame(message.flag(), message);
        }
    }

    protected decodeString(reader: CodecReader, buf: ArrayBuffer, maxLen: number): string {
        const bufView = new DataView(buf);
        let bufViewIdx = 0;

        while (true) {
            const c = reader.getByte();

            if (c == 0 && reader.peekByte() == 10) { //x0a:'\n'
                reader.skipBytes(1);
                break;
            }

            if (maxLen > 0 && maxLen <= bufViewIdx) {
                //超界了，空读。必须读，不然协议流会坏掉
            } else {
                bufView.setInt8(bufViewIdx, c);
                bufViewIdx++;
            }
        }


        if (bufViewIdx < 1) {
            return "";
        }

        //这里要加个长度控制
        return StrUtils.bufToStr(buf, 0, bufViewIdx, this._config.getCharset());
    }
}